agent/lib/kontena/websocket_client.rb

```diff
@@ -74,8 +77,16 @@ def rpc_client
     end
 
     def start
-      every(CONNECT_INTERVAL) do
-        connect! if !connected? unless connecting?
+      loop do
+        if !connected? && !connecting?
+          connect!
+          # Backoff connect attempts between un-succesfull trys
+          sleep @backoff_interval
+          backoff_reconnect!
```

Comment:
This feels a bit wrong, it's hard to understand exactly how this reconnect backoff will behave.

The `connect!` -> `connected!` -> `reset_reconnect_backoff` can happen before the sleep completes and this calls `backoff_reconnect!`... It'll probably still mostly work, but I have no idea as to exactly how it will behave.

Reply:
yep, it's really hard to make it 100% accurate as the `connected!` happens "async" from this loops point of view. Based on my local testing this seems to work nicely, but I agree that there is some randomy feel to it.